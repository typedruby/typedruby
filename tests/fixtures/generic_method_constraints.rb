# @typedruby

def f[T : Numeric](T x, T y) => T
  x < x
end

def test1 => Float
  f(2.0, 3)
end

def test2 => String
  f("foo", "bar")
end

def test3 => Numeric
  f(2.0, 3.0)
end
