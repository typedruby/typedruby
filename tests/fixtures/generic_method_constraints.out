info: Typechecking...
error: Could not match types:

        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:3
      3 |  def f[T : Numeric](T x, T y) => T
                     ^^^^^^^ Numeric, with:
        @ (builtin stdlib):###
    ### |    def <(Numeric other) => Boolean; end
                                     ^^^^^^^ Boolean

        - arising from an attempt to match:

        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:3
      3 |  def f[T : Numeric](T x, T y) => T
                                           ^ Numeric, with:
        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:4
      4 |    x < x
             ^^^^^ Boolean

error: Could not match types:

        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:8
      8 |    f(2.0, 3)
               ^^^ Float, with:
        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:8
      8 |    f(2.0, 3)
                    ^ Integer

        - arising from an attempt to match:

        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:3
      3 |  def f[T : Numeric](T x, T y) => T
                                   ^ Float, with:
        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:8
      8 |    f(2.0, 3)
                    ^ Integer
        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:8
      8 |    f(2.0, 3)
             ^^^^^^^^^ in this expression

error: Could not match types:

        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:3
      3 |  def f[T : Numeric](T x, T y) => T
                     ^^^^^^^ Numeric, with:
        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:12
     12 |    f("foo", "bar")
               ^^^^^ String

        - arising from an attempt to match:

        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:3
      3 |  def f[T : Numeric](T x, T y) => T
                     ^^^^^^^ Numeric, with:
        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:3
      3 |  def f[T : Numeric](T x, T y) => T
                 ^ String
        @ __ROOT__/tests/fixtures/generic_method_constraints.rb:3
      3 |  def f[T : Numeric](T x, T y) => T
                 ^^^^^^^^^^^ in this expression
